lightning/src/chain/keysinterface.rs

```diff
@@ -848,7 +883,27 @@ impl KeysManager {
 	/// Note that until the 0.1 release there is no guarantee of backward compatibility between
 	/// versions. Once the library is more fully supported, the docs will be updated to include a
 	/// detailed description of the guarantee.
+	///
+	/// This method cannot be used for nodes that wish to support receiving multi-node or phantom
+	/// payments; [`KeysManager::new_multi_receive`] must be used instead.
+	///
+	/// Switching between this method and [`KeysManager::new_multi_receive`] will invalidate any
+	/// previously issued invoices and attempts to pay previous invoices will fail.
 	pub fn new(seed: &[u8; 32], starting_time_secs: u64, starting_time_nanos: u32) -> Self {
+		Self::new_inner(seed, starting_time_secs, starting_time_nanos, None)
+	}
+
+	/// Similar to [`KeysManager::new`], but allows the node using this `KeysManager` to receive
+	/// phantom node payments. `cross_node_seed` must be the same across all phantom-receiving nodes
+	/// and also the same across restarts, or else payments may fail.
+	///
+	/// Switching between this method and [`KeysManager::new`] will invalidate any
+	/// previously issued invoices and attempts to pay previous invoices will fail.
+	pub fn new_multi_receive(seed: &[u8; 32], starting_time_secs: u64, starting_time_nanos: u32, cross_node_seed: &[u8; 32]) -> Self {
+		Self::new_inner(seed, starting_time_secs, starting_time_nanos, Some(cross_node_seed))
+	}
```

Comment:
The "phantom-ness" of `KeysManager` bleeds a bit through its public interface here. Instead of having separate constructors for whether or not it's supported (with the accompanying `get_phantom_secret`), there could be two separate types, possibly sharing most of the underlying implementation. Seems this would eliminate the need to have two separate constructors with contracts that could be accidentally violated.

Reply:
This would increase the diff decently (though it'd be boilerplate). I wonder if the risk of using the wrong constructor is that much greater than using the wrong struct would be with this approach (maybe missing sth here).

My other question would be whether we'll want to add more features in the future where a similar argument could be made for a separate KeysManager struct. If this may be the case, it may be that we should modify the existing `new` method to take a `Config` struct, rather than having a separate constructor and/or entirely separate struct.

Not opposed to this approach, but wanted to get some thoughts out first to discuss.

Reply:
> This would increase the diff decently (though it'd be boilerplate). I wonder if the risk of using the wrong constructor is that much greater than using the wrong struct would be with this approach (maybe missing sth here).
> 

Yeah, I'd imagine instead we may be able to extract the logic around key derivation into its own abstraction and parameterize `KeysManager` by that. Maybe something akin to a `KeyDerivation` trait? Then there would still be one `KeysManager` that may be parameterized differently, so in a sense two different structs using the same code with reduced boilerplate.

At least that would be the general idea: take some concept and make a trait/struct out of it.

> My other question would be whether we'll want to add more features in the future where a similar argument could be made for a separate KeysManager struct. If this may be the case, it may be that we should modify the existing `new` method to take a `Config` struct, rather than having a separate constructor or entirely separate struct.
> 

I think this would just change the problem from "Did I call the right constructor?" to "Did I call the constructor with the right parameters?".

The type approach perhaps just moves this further to `ChannelManager`: "Did I use the right type parameter?" But at very least it removes a contractural dependency between `ChannelManager` and a _specific implementation_ of one of its type parameters. Instead, the contract is that `ChannelManager` must be parameterized by the same `KeysInterface` type.

The general principal is dependencies lead to complexity, so aim to reduce dependencies whenever possible.

> Not opposed to this approach, but wanted to get some thoughts out first to discuss.

Of course, there is never right answer. :) My feeling is we should strive to move in the direction of reduced complexity. One answer to that is to limit dependencies. But in practice, sometimes you need to go a step backward before you can go two steps forward.

Reply:
I see, yes it would be nice to get some more assurance that the right thing is being used. The only reason it doesn't feel the most natural to me to parameterize `KeysManager` by a trait object is because we derive the keys on `new`, then cache the keys themselves and would have no reason to store/be parameterized by the `KeyDerivation` trait object after that. 

One way to force users to consciously choose would be to add an enum param to the existing `new` method, with variants for single-node-receive and phantom receive. 

Or we could go with the original idea of a whole separate `PhantomKeysManager` struct. I guess one thing to note is that it seems the enforcement of same-`KeysInterface`-concrete-type-parameter may mostly apply to Rust users, since Java/other users don't have to spell out the concrete types ahead iiuc.

Reply:
I don't have a super strong opinion here, but I'd generally much rather take boilerplate than leak information about how internal stuff to users by templating a struct.

Reply:
Moving ahead with the additional `PhantomKeysManager` struct, thinking about it more I can see how it's potentially a bit cleaner